exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/TypeInferenceUtils.java

```diff
@@ -0,0 +1,571 @@
+          //      - SYMBOL, MULTISET, DISTINCT, STRUCTURED, ROW, OTHER, CURSOR, COLUMN_LIST
+          // .put(SqlTypeName.MAP, TypeProtos.MinorType.MAP)
+          // .put(SqlTypeName.ARRAY, TypeProtos.MinorType.LIST)
+          .build();
+
+  /**
+   * Given a Drill's TypeProtos.MinorType, return a Calcite's corresponding SqlTypeName
+   */
+  public static SqlTypeName getCalciteTypeFromDrillType(final TypeProtos.MinorType type) {
+    return DRILL_TO_CALCITE_TYPE_MAPPING.get(type);
+  }
+
+  /**
+   * Given a Calcite's RelDataType, return a Drill's corresponding TypeProtos.MinorType
+   */
+  public static TypeProtos.MinorType getDrillTypeFromCalciteType(final RelDataType relDataType) {
+    final SqlTypeName sqlTypeName = relDataType.getSqlTypeName();
+    TypeProtos.MinorType minorType = CALCITE_TO_DRILL_MAPPING.get(sqlTypeName);
+    if(minorType == null) {
+      minorType = TypeProtos.MinorType.LATE;
+    }
+    return minorType;
+  }
+
+  /**
+   * Give the name and DrillFuncHolder list, return the inference mechanism.
+   */
+  public static SqlReturnTypeInference getDrillSqlReturnTypeInference(
+      final String name,
+      final List<DrillFuncHolder> functions) {
+    switch(name.toUpperCase()) {
+      case "DATE_PART":
+        return DrillDatePartSqlReturnTypeInference.INSTANCE;
+
+      case "SUM":
+        return new DrillSumSqlReturnTypeInference(functions);
+
+      case "COUNT":
+        return DrillCountSqlReturnTypeInference.INSTANCE;
+
+      case "CONCAT":
+        return DrillConcatSqlReturnTypeInference.INSTANCE;
+
+      case "LENGTH":
+        return DrillLengthSqlReturnTypeInference.INSTANCE;
+
+      case "LPAD":
+      case "RPAD":
+      case "LTRIM":
+      case "RTRIM":
+      case "BTRIM":
+        return DrillPadTrimSqlReturnTypeInference.INSTANCE;
+
+      case "CONVERT_TO":
+        return DrillConvertToSqlReturnTypeInference.INSTANCE;
+
+      case "EXTRACT":
+        return DrillExtractSqlReturnTypeInference.INSTANCE;
+
+      case "CAST":
+        return DrillCastSqlReturnTypeInference.INSTANCE;
+
+      case "FLATTEN":
+      case "KVGEN":
+      case "CONVERT_FROM":
+        return DrillDeferToExecSqlReturnTypeInference.INSTANCE;
+
+      default:
+        return new DrillDefaultSqlReturnTypeInference(functions);
+    }
+  }
+
+  private static class DrillDefaultSqlReturnTypeInference implements SqlReturnTypeInference {
+    private final List<DrillFuncHolder> functions;
+
+    public DrillDefaultSqlReturnTypeInference(List<DrillFuncHolder> functions) {
+      this.functions = functions;
+    }
+
+    @Override
+    public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
+      final RelDataTypeFactory factory = opBinding.getTypeFactory();
+      if (functions.isEmpty()) {
+        return factory.createTypeWithNullability(
+            factory.createSqlType(SqlTypeName.ANY),
+            true);
+      }
+
+      boolean allBooleanOutput = true;
```

Comment:
Why do you treat boolean type specially here? Sounds like the logic could be applied to other type as wells.  If all the DrillFuncHolder's return type is  BigInt, should we return BigInt?